### PR TITLE
Closes #525 derivesummary

### LINKED ISSUE #525: Implement subfunction for derive_summary_records
### Idea:
A function which returns just the summary records should be implemented. It should be used in `derive_exposure_params` (replacing the `derive_summary_records()` and `filter()` call in lines 161-167) and in `derive_summary_records`.

**Details:**
Create a new subfunction which should contain all code of derive_summary_records() except the step where the new records are added to the input dataset. I.e., the subfunction returns only the new records. In derive_summary_records() this new subfunction would be called and thereafter bind_rows() to add the new records to the input dataset.
In derive_params_exposure() the derive_summary_records() call and the subsequent filter() call are replaced by calling the new subfunction.

### Pseudo code @bundfussr please check
add_data <- subset_ds %>%
filter(PARAMCD == input_code) %>%
derive_summary_records(
by_vars = by_vars,
analysis_var = !!analysis_var,
summary_fun = summary_fun,
set_values_to = set_values_to
) %>%
filter(PARAMCD == quo_get_expr(set_values_to$PARAMCD))

### PR DESCRIPTION
**Pull request by alanaharris22**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/) 
- [x] Updated relevant unit tests or have written new unit tests - See [Unit Test Guide](https://pharmaverse.github.io/admiral/articles/unit_test_guidance.html#writing-unit-tests-in-admiral-)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiral/articles/programming_strategy.html#deprecation-1)?
- [x] Update to all relevant roxygen headers and examples 
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md`
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue so that it closes after successful merging. 
- [x] Address all merge conflicts and resolve appropriately 
- [x] Pat yourself on the back for a job well done!  Much love to your accomplishment!

### CODE REVIEW

**Comment by alanaharris22** [comment]
Hi @bms63, 
I'm still getting some errors with the filtering. I am looking into it now (have just added a new commit). I am probably missing something obvious, will keep you posted.

**Comment by alanaharris22** [comment]
Hi @bundfussr, 
The filtering seems to working fine now. Although there seems to be an issue with how `analysis_var` is called - would you mind taking a look? When running an example call of `derive_param_exposure` I get an error stating _'Quosures can only be unquoted within a quasiquotation context.'_, which goes back to `rlang:::Ops.quosure(analysis_var)` when using `last_trace()`. 
Thanks, 
Alana

**Comment by bundfussr** [comment]
Hi @alanaharris22 , which example fails? I tried the examples on the header of the function but they failed with a different error.

**Comment by alanaharris22** [comment]
Hi @bundfussr, 
I used the example from line 76 to 105 in `derive_param_exposure()`, although since merging with devel I am now also getting a different error due to `list2()` - is this what you are getting too? 
Thanks, 
Alana

**Comment by alanaharris22** [comment]
Hi @bundfussr,
 
Sorry for the delay, the code is now working and is ready for review.

For reference I had to add a step in `derive_exposure_param()` to add the original variables in the input dataset to the dataset created by `get_summary_records`, since the code further down expected that these were present. Let me know if you think there's a better solution. 

I'm not sure why the `lint` check is failing though, if I run it directly in R it works fine? It seems there's an issue installing it, do you know what could be the problem?

Thanks, 
Alana

**Comment by bundfussr** [comment]
It seems that a new version of lintr was released recently. The new version does not work with the version of packages we use for development.

@thomas-neitmann , I would suggest that we update the lintr workflow such that it does not use the most recent version of lintr but the one from `renv.lock`. This should make the workflow more stable and any findings from the workflow could be reproduced locally.

What do you think?

**Comment by thomas-neitmann** [comment]
@bundfussr Let's do that! Feel free to approve this PR and ignore the failing lint workflow. I can go ahead and merge using admin privileges.

**Review by bundfussr — CHANGES_REQUESTED** [review]

**Comment by bms63** [comment]
Congrats on making the first update for `admiral 0.8.0` or at least `admiral > 0.7.0` !!   :)

File: NEWS.md [hunk 1]
```diff
@@ -7,6 +7,10 @@ admiral extension packages (#1079).
 
 ## Updates of Existing Functions
 
+- Updates to `derive_summary_records()` and `derive_param_exposure()` to use the new sub function `get_summary_records()` (#525):
+
```
[the remark this answers is not in the document]
Comment (alanaharris22) [reply NEWS.md:11]: Sure, will update.

File: NEWS.md [hunk 2]
```diff
@@ -7,6 +7,10 @@ admiral extension packages (#1079).
...
+  - `get_summary_records()` - creates summary records e.g. derive analysis value (`AVAL`) from multiple records, only      keeping the derived observations
+
```
[the remark this answers is not in the document]
Comment (alanaharris22) [reply NEWS.md:13]: Sure, will update.

File: R/derive_param_exposure.R [hunk 3]
```diff
@@ -161,14 +161,14 @@ derive_param_exposure <- function(dataset,
...
-    derive_summary_records(
+    get_summary_records(
       by_vars = by_vars,
+      filter = PARAMCD == input_code,
       analysis_var = !!analysis_var,
       summary_fun = summary_fun,
       set_values_to = set_values_to
     ) %>%
-    filter(PARAMCD == quo_get_expr(set_values_to$PARAMCD))
+    bind_rows(., dataset[0, ])
```
[the remark this answers is not in the document]
Comment (alanaharris22) [reply R/derive_param_exposure.R:171]: If I didn't have this here I found that the next section of code (lines 173 - 209) then failed as it expected the `add_data` dataset to have ASTDTM and AENDTM variables already.

File: R/get_summary_records.R [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+#'   Set a list of variables to some specified value for the new observation(s)
+#'   + LHS refer to a variable.
+#'   + RHS refers to the values to set to the variable. This can be a string, a symbol, a numeric
+#'   value or NA.
+#'   (e.g.  `vars(PARAMCD = "TDOSE",PARCAT1 = "OVERALL")`).
+#'   More general expression are not allowed.
+#'
+#' @author Pavan Kumar, updated by Alana Harris
+#'
+#' @return A data frame with derived records appended to original dataset.
```
[the remark this answers is not in the document]
Comment (alanaharris22) [reply R/get_summary_records.R:51]: Yes, will remove this.

File: R/get_summary_records.R [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+#' @description
+#' It is not uncommon to have an analysis need whereby one needs to derive an
+#' analysis value (`AVAL`) from multiple records. The ADaM basic dataset
+#' structure variable `DTYPE` is available to indicate when a new derived
+#' records has been added to a dataset.
+#'
+#' @details
+#' When all records have same values within `by_vars` then this function will
+#' retain those common values in the newly derived records. Otherwise new value
+#' will be set to `NA`.
```
[the remark this answers is not in the document]
Comment (alanaharris22) [reply R/get_summary_records.R:12]: Yep, will remove this.

File: R/get_summary_records.R [hunk 6]
```diff
@@ -0,0 +1,170 @@
+#' Create Summary Records
+#'
+#' @description
+#' It is not uncommon to have an analysis need whereby one needs to derive an
+#' analysis value (`AVAL`) from multiple records. The ADaM basic dataset
+#' structure variable `DTYPE` is available to indicate when a new derived
+#' records has been added to a dataset.
```
[the remark this answers is not in the document]
Comment (alanaharris22) [reply R/get_summary_records.R:7]: Sure, will add.

File: R/get_summary_records.R [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+#'   value or NA.
+#'   (e.g.  `vars(PARAMCD = "TDOSE",PARCAT1 = "OVERALL")`).
+#'   More general expression are not allowed.
+#'
+#' @author Pavan Kumar, updated by Alana Harris
+#'
+#' @return A data frame with derived records appended to original dataset.
+#'
+#' @keywords adam user_utility.
+#'
```
[the remark this answers is not in the document]
Comment (alanaharris22) [reply R/get_summary_records.R:54]: Sure, will do.

File: R/derive_param_exposure.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_param_exposure.R:171]: I see. I guess
```
      mutate(
        ASTDTM = coalesce(ASTDTM, temp_start),
        AENDTM = coalesce(AENDTM, temp_end)
      ) %>%
```
is causing the failure.
However, why do we need `coalesce()`? `ASTDTM` and `AENDTM` are NA for all observations. So couldn't we simplify it to 
```
      mutate(
        ASTDTM = temp_start,
        AENDTM = temp_end
      ) %>%
```
and remove ` bind_rows(., dataset[0, ])`?

**Comment by alanaharris22** [comment]
@bundfussr - this PR is now ready to review, sorry for only getting to this now. Not sure why the check_templates isn't running though? 
Thanks!

**Comment by thomas-neitmann** [comment]
"Check Templates" only runs once aomeone approves the PR which I just did 🙂

**Review by thomas-neitmann — APPROVED** [review]

**Comment by bms63** [comment]
> Hi @thomas-neitmann - it looks like check_templates is failing for ADLB, do you know if this is this an problem in other issues too? ad_adlb.R seems to run okay when I run in my R session. Thanks!

we jsut fixed this in another pr.  please merge idevel and you shoudl be good to go

**Comment by alanaharris22** [comment]
> > Hi @thomas-neitmann - it looks like check_templates is failing for ADLB, do you know if this is this an problem in other issues too? ad_adlb.R seems to run okay when I run in my R session. Thanks!
> 
> we jsut fixed this in another pr. please merge idevel and you shoudl be good to go

Thanks! Will do :)

**Comment by alanaharris22** [comment]
@thomas-neitmann @bms63 - would you mind approving the PR so check_templates runs? Thank you!

**Review by thomas-neitmann — APPROVED** [review]

**Comment by thomas-neitmann** [comment]
Done.

**Comment by bms63** [comment]
@thomas-neitmann  Can we merge this one into devel?

**Review by bms63 — APPROVED** [review]
Approving for templates

**Comment by thomas-neitmann** [comment]
@bundfussr Please approve as well. Currently merging is blocked due to your previous ask for changes.

**Review by bundfussr — APPROVED** [review]

**Pull request merged**